x-pack/plugins/infra/public/alerting/inventory/components/expression_chart.tsx

```diff
@@ -0,0 +1,387 @@
+        .valueOf(),
+      to: moment().valueOf(),
+      forceInterval: true,
+      ignoreLookback: true,
+    }),
+    [expression.timeSize, expression.timeUnit]
+  );
+
+  const buildCustomMetric = (metric: any) => ({
+    ...metric,
+    type: 'custom' as SnapshotMetricType,
+  });
+
+  const { loading, nodes } = useSnapshot(
+    filterQuery,
+    expression.metric === 'custom'
+      ? [buildCustomMetric(expression.customMetric)]
+      : [{ type: expression.metric }],
+    [],
+    nodeType,
+    sourceId,
+    0,
+    '',
+    '',
+    true,
+    timerange
+  );
+
+  const { uiSettings } = useKibanaContextForPlugin().services;
+
+  const metric = {
+    field: expression.metric,
+    aggregation: 'avg' as MetricsExplorerAggregation,
+    color: Color.color0,
+  };
+  const isDarkMode = uiSettings?.get('theme:darkMode') || false;
+  const dateFormatter = useMemo(() => {
+    const firstSeries = nodes[0]?.metrics[0]?.timeseries;
+    const firstTimestamp = first(firstSeries?.rows)?.timestamp;
+    const lastTimestamp = last(firstSeries?.rows)?.timestamp;
+
+    if (firstTimestamp == null || lastTimestamp == null) {
+      return (value: number) => `${value}`;
+    }
+
+    return niceTimeFormatter([firstTimestamp, lastTimestamp]);
+  }, [nodes]);
+
+  /* eslint-disable-next-line react-hooks/exhaustive-deps */
+  const yAxisFormater = useCallback(
+    createInventoryMetricFormatter(
+      expression.metric === 'custom'
+        ? buildCustomMetric(expression.customMetric)
+        : { type: expression.metric }
+    ),
+    [expression.metric]
+  );
+
+  if (loading || !nodes) {
+    return (
+      <EmptyContainer>
+        <EuiText color="subdued">
+          <FormattedMessage
+            id="xpack.infra.metrics.alerts.loadingMessage"
+            defaultMessage="Loading"
+          />
+        </EuiText>
+      </EmptyContainer>
+    );
+  }
+
+  const thresholds = expression.threshold.slice().sort();
+
+  // Creating a custom series where the ID is changed to 0
+  // so that we can get a proper domian
+  const firstSeries = nodes[0]?.metrics[0]?.timeseries;
+  if (!firstSeries || !firstSeries.rows || firstSeries.rows.length === 0) {
+    return (
+      <EmptyContainer>
+        <EuiText color="subdued" data-test-subj="noChartData">
+          <FormattedMessage
+            id="xpack.infra.metrics.alerts.noDataMessage"
+            defaultMessage="Oops, no chart data available"
+          />
+        </EuiText>
+      </EmptyContainer>
+    );
+  }
+
+  const thresholdFormatter = (value: number | undefined) => {
+    const metricFormatter = METRIC_FORMATTERS[expression.metric];
+
+    if (!value || !metricFormatter) {
+      return value;
+    }
+
+    switch (metricFormatter.formatter) {
+      case InfraFormatterType.percent:
+        return value / 100;
+      case InfraFormatterType.bits:
+        return value / 1000;
+      default:
+        return value;
+    }
+  };
```

Comment:
We pass a formatter function to the chart component. Threshold values are not in the same unit as the data returned from the snapshot endpoint, so when we render them inside the chart, the formatted function is also applied to thresholds. Is there any better way to handle this?

Reply:
Yeah... it's a little gross. This is being done because we want the user to work with the data in the formatted way. For example, instead of the user typing `0.1` for the threshold, we want them to type `10` for 10 percent, since that's how we display it in the chart. I would expect that the row data to always stay in the original format, then we just use a formatter in the chart to display it correctly and something to translate the percents back to numbers from the inputs.

Reply:
I've been investigating a bit more since for Metric Threshold alerts the preview chart doesn't have to handle this. I saw that In Metric Threshold alerts the value of the threshold is converted after the user enters the value [here](https://github.com/elastic/kibana/blob/31041af43ca368a009ab22d80b91757d2c4df6ed/x-pack/plugins/infra/public/alerting/metric_threshold/components/expression_row.tsx#L139), so internally it's always in the same unit as the chart data.

Would it make sense to do the same for Inventory alerts [here](https://github.com/elastic/kibana/blob/31041af43ca368a009ab22d80b91757d2c4df6ed/x-pack/plugins/infra/public/alerting/inventory/components/expression.tsx#L488)? I guess that I should have to change something else before the alert is saved, although I'm not sure yet where.

Reply:
Yes

Reply:
As discussed in slack, we can't change in which unit thresholds are save since we would have to update the alerts created until now.
So I changed the implementation to convert the thresholds just before they are used.